src/identity/interaction/email-password/credentials/CreateCredentialsHandler.ts

```diff
@@ -0,0 +1,53 @@
+import { randomBytes } from 'crypto';
+import { v4 } from 'uuid';
+import { BasicRepresentation } from '../../../../http/representation/BasicRepresentation';
+import type { Representation } from '../../../../http/representation/Representation';
+import type { KeyValueStorage } from '../../../../storage/keyvalue/KeyValueStorage';
+import { APPLICATION_JSON } from '../../../../util/ContentTypes';
+import { BadRequestHttpError } from '../../../../util/errors/BadRequestHttpError';
+import { NotImplementedHttpError } from '../../../../util/errors/NotImplementedHttpError';
+import type { AccountStore } from '../storage/AccountStore';
+import type { ClientCredentials } from './ClientCredentialsAdapterFactory';
+import type { CredentialsHandlerInput } from './CredentialsHandler';
+import { CredentialsHandler } from './CredentialsHandler';
+
+/**
+ * Handles the creation of credential tokens.
+ * Requires a `name` field in the input JSON body,
+ * that will be used to generate the ID token.
+ */
+export class CreateCredentialsHandler extends CredentialsHandler {
+  private readonly accountStore: AccountStore;
+  private readonly credentialStorage: KeyValueStorage<string, ClientCredentials>;
+
+  public constructor(accountStore: AccountStore, credentialStorage: KeyValueStorage<string, ClientCredentials>) {
+    super();
+    this.accountStore = accountStore;
+    this.credentialStorage = credentialStorage;
+  }
+
+  public async canHandle({ body }: CredentialsHandlerInput): Promise<void> {
+    if (typeof body.name !== 'string') {
+      throw new NotImplementedHttpError();
+    }
+  }
+
+  public async handle({ operation, body: { webId, name }}: CredentialsHandlerInput): Promise<Representation> {
+    const settings = await this.accountStore.getSettings(webId);
+
+    if (!settings.useIdp) {
+      throw new BadRequestHttpError('This server is not an identity provider for this account.');
+    }
+
+    const id = `${(name as string).replace(/\W/gu, '-')}_${v4()}`;
+    const secret = randomBytes(64).toString('hex');
+
+    await this.credentialStorage.set(id, { secret, webId });
+    settings.clientCredentials = settings.clientCredentials ?? [];
+    settings.clientCredentials.push(id);
+    await this.accountStore.updateSettings(webId, settings);
```

Comment:
I suppose in case of failure of the account store, the dangling credentials are not a big issue.

Reply:
It would be if someone would manage to somehow find out the id/secret combination. I'll reverse the order since having a non-existent token in your settings is less of an issue.